### ISSUE TITLE
Horovod accelerator using full dataset on each worker.

### ISSUE THREAD

**Issue by jipson7** [PyTorchLightning/pytorch-lightning#8773]
## 🐛 Bug

Hi there. I have been successfully using pytorch lightning with the ddp accelerator on a multi-gpu setup for a while. I wanted to test horovod and noticed it is an order of magnitude slower. I found the problem to be that each worker is processing all the data in an epoch, where as normally with DDP each GPU would have an equal subset of the dataset.

Im forced to launch the job with mpirun due to it being a "managed" gpu environment. Could this cause the issue? 

I cannot upload the code due to IP issues, and I have been trying to recreate an minimal example but Im not sure where the issue is.

Sorry for the vague post and thanks for any help.

### Expected behavior

I.E. if my dataset has 100 items with a batch size of 1, on a 4 gpu setup, with ddp each GPU would see 25 samples, with horovod each GPU is seeing 100 samples. Im not sure if the gradients are being merged either as it doesnt seem to be training any faster. Expected outcome is an equal subset of the data on each worker and faster convergence.

### Environment

* CUDA:
	- GPU:
		- Tesla V100-SXM2-32GB
		- Tesla V100-SXM2-32GB
		- Tesla V100-SXM2-32GB
		- Tesla V100-SXM2-32GB
		- Tesla V100-SXM2-32GB
		- Tesla V100-SXM2-32GB
		- Tesla V100-SXM2-32GB
		- Tesla V100-SXM2-32GB
	- available:         True
	- version:           11.1
* Packages:
	- numpy:             1.20.3
	- pyTorch_debug:     False
	- pyTorch_version:   1.8.1+cu111
	- pytorch-lightning: 1.4.0
	- tqdm:              4.62.0
* System:
	- OS:                Linux
	- architecture:
		- 64bit
		- ELF
	- processor:         x86_64
	- python:            3.8.9
	- version:           #1 SMP Mon Jul 29 17:46:05 UTC 2019

**Comment by awaelchli** [comment]
Hey @jipson7 
Could you check if Lightning added a distributed sampler automatically and that horovod accelerator is selected?
For example, inside the LightningModule training step:

```python
print(self.trainer.train_dataloader.sampler)
print(self.trainer.training_type_plugin)
```

**Comment by jipson7** [comment]
Hi @awaelchli thanks for the pointer. Here's the output.

<torch.utils.data.distributed.DistributedSampler object at 0x7f48323caee0>
<pytorch_lightning.plugins.training_type.ddp.DDPPlugin object at 0x7f483e916a90>

Does this mean the Horovod accelerator is not selected? For a bit more context Im using the yaml files provided with the LightningCLI and setting accelerator to `horovod`

**Comment by awaelchli** [comment]
Yes, it's not selecting the horovod accelerator. Is it possible to reproduce it with our mini model here https://github.com/PyTorchLightning/pytorch-lightning/blob/master/pl_examples/bug_report_model.py
by setting the trainer arguments manually there? And can you share the launch command? 

My understanding is that it can work with mpirun or horovodrun

**Comment by jipson7** [comment]
I was using the below code snippet with the Lightning CLI because I couldnt get `find_unused_parameters` to set properly in the config file. Removing it solved the issue. Thank you.

```
class MyLightningCLI(LightningCLI):
    def add_arguments_to_parser(self, parser):
        parser.set_defaults({'trainer.plugins': pl.plugins.DDPPlugin(find_unused_parameters=True)})
```
